### PR TITLE
Implement -pingboost 4 for behavior similar to mm_insane 1 in Linux

### PR DESCRIPTION
**Pull request by cris840**
Implemented the Sleep_Pingboost4 function to achieve similar behavior to mm_insane 1 of the mmtimer metamod, since the function does not perform any further action, sleep is removed.

### CODE REVIEW

File: rehlds/dedicated/src/sys_linux.cpp [hunk 1]
```diff
@@ -146,34 +146,40 @@ void alarmFunc(int num)
...
+        case 1:
+            signal(SIGALRM, alarmFunc);
+            Sys_Sleep = Sleep_Timer;
+            break;
+        case 2:
+            Sys_Sleep = Sleep_Select;
+            break;
+        case 3:
+            Sys_Sleep = Sleep_Net;
+            NET_Sleep_Timeout = (NET_Sleep_t)Sys_GetProcAddress(g_pEngineModule, "NET_Sleep_Timeout");
```
Comment (justgo97) [inline rehlds/dedicated/src/sys_linux.cpp:168]: Bring back the comments

File: rehlds/dedicated/src/sys_linux.cpp [hunk 2]
```diff
@@ -146,34 +146,40 @@ void alarmFunc(int num)
...
+            Sys_Sleep = Sleep_Pingboost4;
+            break;
+        default:
+            Sys_Sleep = Sleep_Old;
+            break;
+        }
+    }
+}
+
+void Sleep_Pingboost4(int msec)
```
Comment (justgo97) [inline rehlds/dedicated/src/sys_linux.cpp:180]: Move this above so you can use it directly

**Review by cris840 — COMMENTED** [review]
Alright

**Comment by axl303** [comment]
Sorry for my stupid comment/question,
What would be the benefits/is there any from -pingboost4 and should we use it over -pingboost3?

**Comment by cris840** [comment]
> Sorry for my stupid comment/question, What would be the benefits/is there any from -pingboost4 and should we use it over -pingboost3?

It's simple, in short:

-pingboost 3 is designed to control Sleep on network operations with precision while the network subsystem does its internal tasks.
 
-pingboost 4 does not perform any Sleep and simply returns the function without doing anything. This means that the CPU will be fully utilized without waiting or suspending execution. This can be used in order to increase FPS to "insane" levels ;)
Like mmtimer Metamod

**Comment by SmilexGamer** [comment]
> Implemented the Sleep_Pingboost4 function to achieve similar behavior to mm_insane 1 of the mmtimer metamod, since the function does not perform any further action, sleep is removed.

Couldn't this be implemented also in Windows?

**Comment by cris840** [comment]
> > Implemented the Sleep_Pingboost4 function to achieve similar behavior to mm_insane 1 of the mmtimer metamod, since the function does not perform any further action, sleep is removed.
> 
> Couldn't this be implemented also in Windows?

On Windows you can use the mmtimer metamod directly and use mm_insane 1 in server.cfg to have the same result.

**Comment by SmilexGamer** [comment]
> > > Implemented the Sleep_Pingboost4 function to achieve similar behavior to mm_insane 1 of the mmtimer metamod, since the function does not perform any further action, sleep is removed.
> > 
> > 
> > Couldn't this be implemented also in Windows?
> 
> On Windows you can use the mmtimer metamod directly and use mm_insane 1 in server.cfg to have the same result.

Doesn't that apply also to Linux? You can use the mmtimer too. If we're implementing this in reHLDS, it should be in both platforms.

**Comment by cris840** [comment]
> 



> > > > Implementé la función Sleep_Pingboost4 para lograr un comportamiento similar a mm_insane 1 del metamod mmtimer, ya que la función no realiza ninguna otra acción, se elimina el sueño.
> > > 
> > > 
> > > ¿No se podría implementar esto también en Windows?
> > 
> > 
> > En Windows, puede usar el metamod mmtimer directamente y usar mm_insane 1 en server.cfg para obtener el mismo resultado.
> 
> ¿Eso no se aplica también a Linux? También puede utilizar el mmtimer. Si estamos implementando esto en reHLDS, debería estar en ambas plataformas.



> > > > Implemented the Sleep_Pingboost4 function to achieve similar behavior to mm_insane 1 of the mmtimer metamod, since the function does not perform any further action, sleep is removed.
> > > 
> > > 
> > > Couldn't this be implemented also in Windows?
> > 
> > 
> > On Windows you can use the mmtimer metamod directly and use mm_insane 1 in server.cfg to have the same result.
> 
> Doesn't that apply also to Linux? You can use the mmtimer too. If we're implementing this in reHLDS, it should be in both platforms.

pingboost does not exist in the Windows version of ReHLDS
Look: https://github.com/dreamstalker/rehlds/blob/master/rehlds/dedicated/src/sys_window.cpp

**Comment by cris840** [comment]
> > > > Implemented the Sleep_Pingboost4 function to achieve similar behavior to mm_insane 1 of the mmtimer metamod, since the function does not perform any further action, sleep is removed.
> > > 
> > > 
> > > Couldn't this be implemented also in Windows?
> > 
> > 
> > On Windows you can use the mmtimer metamod directly and use mm_insane 1 in server.cfg to have the same result.
> 
> Doesn't that apply also to Linux? You can use the mmtimer too. If we're implementing this in reHLDS, it should be in both platforms.

And no, unfortunately there is no working version of mmtimer for Linux. It would need to be updated, but there is no point in doing so when it can be implemented directly in ReHLDS.

**Comment by stambeto2006** [comment]
@cris840 It would be a good idea and useful if it was built to work on Windows as well without adding unnecessary extras that are not guaranteed to work 100%.

**Comment by hun1er** [comment]
@stambeto2006 , @SmilexGamer You can use [this](https://github.com/hun1er/rehlds-m/releases/tag/v1.3.0) which is compatible with HLDS and ReHLDS and works on both Windows and Linux. It has a similar feature (pingboost 5) and even more.

**Comment by rtxa** [comment]
@hun1er Wouldn't be better to do a PR with your changes instead of using your fork? Otherwise it always get behinds with latest ReHLDS.

**Comment by anzz1** [comment]
This doesn't really make any sense as the max packet rate is 100/s anyway and already with the original method the server's tickrate is way above that.
What this does is completely unnecessarily waste CPU cycles and energy and could cause server providers to not adopt ReHLDS or even outright ban it.

**Comment by SmilexGamer** [comment]
> Yes of course it makes difference, to a point. This however, makes no sense. As long as the server can process the frame in time for the next packet, the rest is pointless. In fact the absolute minimum frametime is 1ms anyway so anything beyond that will do absolutely nothing but waste cycles.

That's odd, since I've seen a difference past 1ms frametime. Feel free to do the same test using ReHLDS-M.

**Pull request closed without merging**